### ISSUE TITLE
Propagate windows service start parameters

### ISSUE THREAD

**Issue by 2opremio** [kardianos/service#82]
This library [simply ignores and doesn't propagate the start parameters of windows services](https://github.com/kardianos/service/blob/5ab7ce2c8f38cc0ca92d711f7c4f3ea70d347702/service_windows.go#L158). 

Those live parameters are very convenient and flexible, allow to avoid configuration files or registry keys to store the parameters.

Related to #75

**Comment by 2opremio** [comment]
A new method (e.g. `Args() []string`) could be added to [`Service`](https://godoc.org/github.com/kardianos/service#Service) without breaking backwards compatibility.

**Comment by 2opremio** [comment]
The `Start()` method should also pass the arguments (and that one is tricky to implement without breaking backwards compatibility)

**Comment by kardianos** [comment]
Are the args passed into Execute different then os.Args()?

**Comment by 2opremio** [comment]
Yes, Windows seems to distinguish between:

* The arguments (`os.Args()`) passed to the Service Application binary managing services
* The arguments (provided by `windows/svc.Execute()`) passed to a service on start.

AFAIU the arguments passed to `Execute()` are the ones provided to [StartProcess](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686321(v=vs.85).aspx) (from `advapi.dll`)

As an example, both `sc.exe` and the Services UI let's you pass those arguments on start:

```
$ sc.exe start
DESCRIPTION:
        Starts a service running.
USAGE:
        sc <server> start [service name] <arg1> <arg2> ...
The following WinRM command responded with a non-zero exit status.
```

<img width="1772" alt="screen shot 2017-02-20 at 19 46 30" src="https://cloud.githubusercontent.com/assets/2362916/23138353/42fda646-f7a7-11e6-9b4c-ee627fd3526a.png">

Note that, in the last example you are free to choose start arguments different to the ones passed to the binary (`-k netsvcs`)

**Comment by zhengxiaoyao0716** [comment]
@kardianos I got the problem too.
First, when I run my code with `go run`, it works correctly:
![image](https://user-images.githubusercontent.com/12682063/28661783-6dd0c650-72ea-11e7-81ec-daed8a12ee9d.png)
The first two line print by `fmt`, execute before service start.
The other print by `log`, witch would redirect to log file in service mode.

However, when I run my code with builded binary executer, the problem appeared:
![image](https://user-images.githubusercontent.com/12682063/28661936-04c77482-72eb-11e7-8be8-f3c243edd822.png)
![image](https://user-images.githubusercontent.com/12682063/28662095-82a7f44e-72eb-11e7-8549-ba48b1ae6b85.png)
You can see that `fmt` print the `os.Args` correctly, while `log` missing them.

**Comment by rgl** [comment]
Any news about this? Something like what @2opremio proposed would be really nice to have :-)

**Comment by rgl** [comment]
Oh, we can already manually propagate arguments! Sorry about the noise...

I ended up using something like:

```go
	nameFlag := flag.String("name", "dummy", "Service name.")
	flag.Parse()
	s, err := service.New(
		&dummyService{},
		&service.Config{
			Arguments:   []string{"-name", *nameFlag},
			Name:        *nameFlag,
			DisplayName: "Dummy Windows Service",
			Description: "Dummy Windows Service",
		})

```

**Comment by kardianos** [comment]
Yes.

**Comment by chuanbozhang-okta** [comment]
> Oh, we can already manually propagate arguments! Sorry about the noise...
> 		&service.Config{
> 			Arguments:   []string{"-name", *nameFlag},
> 		})
> ```
@rgl , does this example actually propagate the live argument as @2opremio called out?
And how to access this parameters from the actual service start method when the parameter not passed through executable but through sc.exe or service control manager?

**Comment by chuanbozhang-okta** [comment]
I did more investigation, and see adding a line of "ws.Arguments = append(ws.Arguments, args...)" in Execute method of service_windows.go solved the issue for me.
```
func (ws *windowsService) Execute(args []string, r <-chan svc.ChangeRequest, changes chan<- svc.Status) (bool, uint32) {
	const cmdsAccepted = svc.AcceptStop | svc.AcceptShutdown
	changes <- svc.Status{State: svc.StartPending}
	ws.Arguments = append(ws.Arguments, args...)	
```

then in my usage, I declare service.Config to my program struct, and then in program Start method access the parameter from p.config.Arguments, and I've validated that could propagate the live arguments from service control manager.

@kardianos , how do you feel this solution?

**Comment by msays2000** [comment]
@chuanbozhang-okta @kardianos @rgl Is support for runtime arguments like below supported in service start method?

`sc start serviceName arg1 arg2 ... argN`

**Comment by chuanbozhang-okta** [comment]
@msays2000 , yes, my proposal fix above to add "ws.Arguments = append(ws.Arguments, args...)" to Execute method could fix the issue and make your usage work.

@kardianos , do you accept contribution for this? If so I could send a PR for it.

**Comment by msays2000** [comment]
@chuanbozhang-okta thank you for confirming. 
@kardianos please review and help in merging the above PR of @chuanbozhang-okta 

We have a use case where we need to start a grpc service on a determined port passed to the service via its start arguments.

**Comment by msays2000** [comment]
@kardianos is there an alternative to pass the arguments to a service.
